src/app/navigation/types.ts

```diff
@@ -15,24 +15,12 @@ export type RouteState<R> = R extends Route<any, infer S> ? S : never;
 type UnionRouteMatches<R> = R extends AnyRoute ? Match<R> : never;
 type UnionHistoryActions<R> = R extends AnyRoute ? HistoryAction<R> : never;
 
-export interface BasicMatch<R> {
-  route: R;
-}
-export interface MatchWithParams<R extends AnyRoute> extends BasicMatch<R> {
-  params: RouteParams<R>;
-}
-export interface MatchWithState<R extends AnyRoute> extends BasicMatch<R> {
-  state?: RouteState<R>;
-}
-
-export type Match<R> = R extends AnyRoute
-  ?
-    (RouteParams<R> extends undefined
-      ? BasicMatch<R>
-      : MatchWithParams<R>)
-    & (RouteState<R> extends undefined
-      ? BasicMatch<R>
-      : MatchWithState<R>)
+export type Match<R> = R extends Route<infer P, infer S>
+  ? {
+    route: R,
+    params: P,
+    state: S,
+  }
```

Comment:
its less great because you have these params defined when they're not used for some routes, but the code is much simpler with less switches so i think it mostly just offends my human brain